core/common/src/main/java/alluxio/worker/block/io/TimeBoundBlockWriter.java

```diff
@@ -0,0 +1,116 @@
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.worker.block.io;
+
+import alluxio.metrics.MetricKey;
+import alluxio.metrics.MetricsSystem;
+import alluxio.network.protocol.databuffer.DataBuffer;
+
+import com.codahale.metrics.Counter;
+import com.google.common.base.Throwables;
+import com.google.common.util.concurrent.SimpleTimeLimiter;
+import com.google.common.util.concurrent.TimeLimiter;
+import io.netty.buffer.ByteBuf;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.WritableByteChannel;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * A wrapper class on {@link BlockWriter} with timeout.
+ * Note that, this reader will not queue any request.
+ */
+public class TimeBoundBlockWriter extends BlockWriter {
+  private final BlockWriter mBlockWriter;
+  private final long mTimeoutMs;
+  private final TimeLimiter mTimeLimter;
+  private final ExecutorService mExecutorService;
+
+  /**
+   * Creates new time bounded block writer.
+   *
+   * @param blockWriter the block writer
+   * @param timeoutMs timeout in milliseconds
+   * @param timeoutThreads threads to execute read operations and timeout
+   */
+  public TimeBoundBlockWriter(BlockWriter blockWriter, long timeoutMs, int timeoutThreads)
+      throws IOException {
+    mBlockWriter = blockWriter;
+    mTimeoutMs = timeoutMs;
+    mExecutorService = new ThreadPoolExecutor(timeoutThreads,
+        timeoutThreads, 0L, TimeUnit.MILLISECONDS, new SynchronousQueue<>());
+    mTimeLimter = SimpleTimeLimiter.create(mExecutorService);
+  }
+
+  @Override
+  public long append(ByteBuffer inputBuf) throws IOException {
+    return callWithTimeout(() -> mBlockWriter.append(inputBuf));
+  }
+
+  @Override
+  public long append(ByteBuf buf) throws IOException {
+    return callWithTimeout(() -> mBlockWriter.append(buf));
+  }
+
+  @Override
+  public long append(DataBuffer buffer) throws IOException {
+    return callWithTimeout(() -> mBlockWriter.append(buffer));
+  }
+
+  @Override
+  public long getPosition() {
+    return mBlockWriter.getPosition();
+  }
+
+  @Override
+  public WritableByteChannel getChannel() {
+    return mBlockWriter.getChannel();
+  }
+
+  @Override
+  public void close() throws IOException {
+    mExecutorService.shutdown();
+    mBlockWriter.close();
+  }
+
+  private <T> T callWithTimeout(Callable<T> callable) throws IOException {
+    try {
+      return mTimeLimter.callWithTimeout(callable, mTimeoutMs, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new IOException(e);
```

Comment:
I throw a runtime exception when the thread is interrupted in local cache's implementation.

Reply:
Change to throw runtime exception
Other places in code base also use runtime exception
IterruptedException in server is non-recoverable